### ISSUE TITLE
App is crashing when dragging custom cell to a new section

### ISSUE THREAD

**Issue by manish026** [adamshin/SwiftReorder#45]
Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'attempt to delete row 2 from section 0 which only contains 2 rows before the update'

Getting this error while dragging 3rd row from 0th section to 1st row of 1st section

**Comment by manish026** [comment]
the fix is call **reorderRowAt** delegate after deleting the row

tableView.beginUpdates()
        
        tableView.deleteRows(at: [context.destinationRow], with: .fade)
        
        delegate?.tableView(tableView, reorderRowAt: context.destinationRow, to: newContext.destinationRow)
        
        tableView.insertRows(at: [newContext.destinationRow], with: .fade)
        tableView.endUpdates()

**Comment by adamshin** [comment]
Are you updating your data model in `tableView(_:UITableView, reorderRowAt:IndexPath, to:IndexPath)`? A couple other people had this issue recently – I left a comment [here](https://github.com/adamshin/SwiftReorder/issues/32#issuecomment-452557776) explaining how to resolve it.

**Comment by manish026** [comment]
Yes I was updating data model in tableView(_:UITableView, reorderRowAt:IndexPath, to:IndexPath)

**Comment by thejeff77** [comment]
I had a similar issue because somewhere in my code I was calling tableView.endUpdates() to fix a bug with TBEmptyDataSet, and had never called tableView.beginUpdates() - you might check that you have matching beginUpdates and endUpdates calls for any updates you have.

**Comment by cliftonlabrum** [comment]
I'm getting this same crash. I tried the suggestion of adding the `delegate?.tableView...` call after deleting the row, but it doesn't work. I've confirmed that I don't have any additional `tableView.beginUpdates()` calls anywhere. 

Anyone else running into this?

**Comment by cliftonlabrum** [comment]
@thejeff77 Thanks. Are you referring to something other than just the built-in UIKit way of reordering `UITableView` rows? My goal is to not have the reorder lines/icon that those (used to?) put on a table row since my cells have a custom look.

Do you have a link to that sample project you are referring to?

**Comment by thejeff77** [comment]
@cliftonlabrum no I am referring to UIKit drag and drop delegates, although there are ways to customize the drag placeholder view. I'm not sure what you mean by the icons/lines but I know of some ways to remove the 1px separators from the cells. I received the example project from apple code support, so I won't post it but I'll try to get around to emailing it to you. The example doesn't have multiple sections tho.

**Comment by cliftonlabrum** [comment]
I found a super helpful Stack Overflow post that shows how easy it is to enable drag and drop for a `UITableView`. If all you were using this library for was reordering a table without the default reorder control, then this is a simple way to get that working: https://stackoverflow.com/questions/60270449/how-can-i-use-drag-and-drop-to-reorder-a-uitableview

Just add the drag delegates to your table:
```
tableView.dragDelegate = self
tableView.dragInteractionEnabled = true
```
...and one delegate method:
```
func tableView(_ tableView: UITableView, itemsForBeginning session: UIDragSession, at indexPath: IndexPath) -> [UIDragItem] {
  let dragItem = UIDragItem(itemProvider: NSItemProvider())
  return [dragItem]
}
```
Then use the usual `moveRowAt` and `canMoveRowAt` methods. It works great! 🙂
